File: cli/src/main/scala/net/kemitix/thorp/cli/CliArgs.scala
package net.kemitix.thorp.cli

import java.nio.file.Paths

import scala.jdk.CollectionConverters._

import net.kemitix.thorp.config.{ConfigOption, ConfigOptions}
import scopt.OParser
import zio.Task

object CliArgs {

  def parse(args: List[String]): Task[ConfigOptions] = Task {
    OParser
      .parse(configParser, args, List())
      .map(options => ConfigOptions(options.asJava))
      .getOrElse(ConfigOptions.empty)
  }

  val configParser: OParser[Unit, List[ConfigOption]] = {
    val parserBuilder = OParser.builder[List[ConfigOption]]
    import parserBuilder._
    OParser.sequence(
      programName("thorp"),
      head("thorp"),
      opt[Unit]('V', "version")
        .action((_, cos) => ConfigOption.version() :: cos)
        .text("Show version"),
      opt[Unit]('B', "batch")
        .action((_, cos) => ConfigOption.batchMode() :: cos)
        .text("Enable batch-mode"),
      opt[String]('s', "source")
        .unbounded()
        .action((str, cos) => ConfigOption.source(Paths.get(str)) :: cos)
        .text("Source directory to sync to destination"),
      opt[String]('b', "bucket")
        .action((str, cos) => ConfigOption.bucket(str) :: cos)
        .text("S3 bucket name"),
      opt[String]('p', "prefix")
        .action((str, cos) => ConfigOption.prefix(str) :: cos)
        .text("Prefix within the S3 Bucket"),
      opt[Int]('P', "parallel")
        .action((int, cos) => ConfigOption.parallel(int) :: cos)
        .text("Maximum Parallel uploads"),
      opt[String]('i', "include")
        .unbounded()
        .action((str, cos) => ConfigOption.include(str) :: cos)
        .text("Include only matching paths"),
      opt[String]('x', "exclude")
        .unbounded()
        .action((str, cos) => ConfigOption.exclude(str) :: cos)
        .text("Exclude matching paths"),
      opt[Unit]('d', "debug")
        .action((_, cos) => ConfigOption.debug() :: cos)
        .text("Enable debug logging"),
      opt[Unit]("no-global")
        .action((_, cos) => ConfigOption.ignoreGlobalOptions() :: cos)
        .text("Ignore global configuration"),
      opt[Unit]("no-user")
        .action((_, cos) => ConfigOption.ignoreUserOptions() :: cos)
        .text("Ignore user configuration")
    )
  }

}

File: cli/src/main/java/net/kemitix/thorp/cli/CliArgs.java
package net.kemitix.thorp.cli;

import net.kemitix.thorp.config.ConfigOption;
import net.kemitix.thorp.config.ConfigOptions;
import picocli.CommandLine;
import picocli.CommandLine.Option;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Collections;
import java.util.List;
import java.util.function.Function;
import java.util.function.Supplier;
import java.util.stream.Collectors;

public class CliArgs {
    @Override
    public String toString() {
        return "CliArgs{" +
                "showVersion=" + showVersion +
                ",\n batchMode=" + batchMode +
                ",\n sources=" + sources +
                ",\n bucket='" + bucket + '\'' +
                ",\n prefix='" + prefix + '\'' +
                ",\n parallel=" + parallel +
                ",\n includes=" + includes +
                ",\n excludes=" + excludes +
                ",\n debug=" + debug +
                ",\n ignoreUserOptions=" + ignoreUserOptions +
                ",\n ignoreGlobalOptions=" + ignoreGlobalOptions +
                "\n}";
    }

    public static ConfigOptions parse(String[] args) {
        CliArgs cliArgs = new CliArgs();
        new CommandLine(cliArgs).parseArgs(args);
        return ConfigOptions.empty()
                .merge(flag(cliArgs.showVersion, ConfigOption::version))
                .merge(flag(cliArgs.debug, ConfigOption::debug))
                .merge(flag(cliArgs.batchMode, ConfigOption::batchMode))
                .merge(flag(cliArgs.ignoreGlobalOptions, ConfigOption::ignoreGlobalOptions))
                .merge(flag(cliArgs.ignoreUserOptions, ConfigOption::ignoreUserOptions))
                .merge(option(cliArgs.bucket, ConfigOption::bucket))
                .merge(option(cliArgs.prefix, ConfigOption::prefix))
                .merge(paths(cliArgs.sources, ConfigOption::source))
                .merge(strings(cliArgs.includes, ConfigOption::include))
                .merge(strings(cliArgs.excludes, ConfigOption::exclude))
                .merge(integer(cliArgs.parallel, 1, ConfigOption::parallel));
    }

    private static ConfigOptions flag(
            boolean flag,
            Supplier<ConfigOption> configOption
    ) {
        if (flag)
            return ConfigOptions.create(Collections.singletonList(
                    configOption.get()));
        return ConfigOptions.empty();
    }

    private static ConfigOptions option(
            String value,
            Function<String, ConfigOption> configOption
    ) {
        if (value.isEmpty())
            return ConfigOptions.empty();
        return ConfigOptions.create(Collections.singletonList(
                configOption.apply(value)));
    }

    private static ConfigOptions integer(
            int value,
            int defaultValue,
            Function<Integer, ConfigOption> configOption
    ) {
        if (value == defaultValue) return ConfigOptions.empty();
        return ConfigOptions.create(Collections.singletonList(
                configOption.apply(value)));
    }

    private static ConfigOptions paths(
            List<String> values,
            Function<Path, ConfigOption> configOption
    ) {
        if (values == null) return ConfigOptions.empty();
        return ConfigOptions.create(
                values.stream()
                        .map(Paths::get)
                        .map(configOption)
                        .collect(Collectors.toList()));
    }

    private static ConfigOptions strings(
            List<String> values,
            Function<String, ConfigOption> configOption
    ) {
        if (values == null) return ConfigOptions.empty();
        return ConfigOptions.create(
                values.stream()
                        .map(configOption)
                        .collect(Collectors.toList()));
    }

    @Option(
            names = {"-V", "--version"},
            description = "Show version")
    boolean showVersion;

    @Option(
            names = {"-B", "--batch"},
            description = "Enable batch-mode")
    boolean batchMode;

    @Option(
            names = {"-s", "--source"},
            description = "Source directory to sync to destination")
    List<String> sources;

    @Option(
            names = {"-b", "--bucket"},
            defaultValue = "",
            description = "S3 bucket name")
    String bucket;

    @Option(
            names = {"-p", "--prefix"},
            defaultValue = "",
            description = "Prefix within the S3 Bucket")
    String prefix;

    @Option(
            names = {"-P", "--parallel"},
            defaultValue = "1",
            description = "Maximum Parallel uploads")
    int parallel;

    @Option(
            names = {"-i", "--include"},
            description = "Include only matching paths")
    List<String> includes;

    @Option(
            names = {"-x", "--exclude"},
            description = "Exclude matching paths")
    List<String> excludes;

    @Option(
            names = {"-d", "--debug"},
            description = "Enable debug logging")
    boolean debug;

    @Option(
            names = {"--no-user"},
            description = "Ignore user configuration")
    boolean ignoreUserOptions;

    @Option(
            names = {"--no-global"},
            description = "Ignore global configuration")
    boolean ignoreGlobalOptions;

}
